File: common/types/game.py
from enum import Enum

from generated import game_pb2


class Type(str, Enum):
    TIC_TAC_TOE = 'tic_tac_toe'
    BATTLESHIPS = 'battleships'


class Result(str, Enum):
    TBD = 'tbd'
    DRAW = 'draw'
    P1_WON = 'p1_won'
    P2_WON = 'p2_won'


class Status(str, Enum):
    IN_QUEUE = 'in_queue'
    IN_PROGRESS = 'in_progress'
    ABANDONED = 'abandoned'
    FINISHED = 'finished'


TYPE_TO_PROTO_MAP: dict[Type, game_pb2.GameType.ValueType] = {
    Type.TIC_TAC_TOE: game_pb2.TIC_TAC_TOE,
    Type.BATTLESHIPS: game_pb2.BATTLESHIPS,
}

RESULT_TO_PROTO_MAP: dict[Result, game_pb2.GameResult.ValueType] = {
    Result.TBD: game_pb2.TBD,
    Result.DRAW: game_pb2.DRAW,
    Result.P1_WON: game_pb2.P1_WON,
    Result.P2_WON: game_pb2.P2_WON,
}

STATUS_TO_PROTO_MAP: dict[Status, game_pb2.GameStatus.ValueType] = {
    Status.IN_QUEUE: game_pb2.IN_QUEUE,
    Status.IN_PROGRESS: game_pb2.IN_PROGRESS,
    Status.ABANDONED: game_pb2.ABANDONED,
    Status.FINISHED: game_pb2.FINISHED,
}

TYPE_FROM_PROTO_MAP: dict[game_pb2.GameType.ValueType, Type] = {
    game_pb2.TIC_TAC_TOE: Type.TIC_TAC_TOE,
    game_pb2.BATTLESHIPS: Type.BATTLESHIPS,
}

RESULT_FROM_PROTO_MAP: dict[game_pb2.GameResult.ValueType, Result] = {
    game_pb2.TBD: Result.TBD,
    game_pb2.DRAW: Result.DRAW,
    game_pb2.P1_WON: Result.P1_WON,
    game_pb2.P2_WON: Result.P2_WON,
}

STATUS_FROM_PROTO_MAP: dict[game_pb2.GameStatus.ValueType, Status] = {
    game_pb2.IN_QUEUE: Status.IN_QUEUE,
    game_pb2.IN_PROGRESS: Status.IN_PROGRESS,
    game_pb2.ABANDONED: Status.ABANDONED,
    game_pb2.FINISHED: Status.FINISHED,
}

File: services/ui/src/features/games/types/game.ts
import { z } from "zod"

type GameType = "tic_tac_toe" | "battleships"

type GameResult = "tbd" | "draw" | "p1_won" | "p2_won"

type GameStatus = "in_queue" | "in_progress" | "abandoned" | "finished"

type Game = {
    id: number
    type: GameType
    result: GameResult
    status: GameStatus
    player1: number
    player2: number | null
    current_player: number
    turn_number: number
    created_at: string
    updated_at: string
}

interface GameCreate {
    type: GameType
}

export type { Game, GameType, GameResult, GameStatus, GameCreate }

export const TokenMessageScheme = z.object({
    type: z.literal("token"),
    body: z.object({ token: z.string() }),
})
export const AuthenticatedMessageScheme = z.object({
    type: z.literal("authenticated"),
    body: z.object({ success: z.boolean(), message: z.string() }),
})
export const GameMoveMessageScheme = z.object({
    type: z.literal("game_move"),
    body: z.object({ game_move: z.record(z.any(), z.any()) }),
})
export const GameStateMessageScheme = z.object({
    type: z.literal("game_state"),
    body: z.object({ game_state: z.unknown() }),
})
export const GameCreatedMessageScheme = z.object({
    type: z.literal("game_created"),
    body: z.object({ message: z.string() }),
})
export const ConnectedMessageScheme = z.object({
    type: z.literal("connected"),
    body: z.object({ message: z.string() }),
})
export const DisconnectedMessageScheme = z.object({
    type: z.literal("disconnected"),
    body: z.object({ message: z.string() }),
})
export const InvalidMoveMessageScheme = z.object({
    type: z.literal("invalid_move"),
    body: z.object({ message: z.string() }),
})

export const GameMessageScheme = z.discriminatedUnion("type", [
    TokenMessageScheme,
    AuthenticatedMessageScheme,
    GameMoveMessageScheme,
    GameStateMessageScheme,
    GameCreatedMessageScheme,
    ConnectedMessageScheme,
    DisconnectedMessageScheme,
    InvalidMoveMessageScheme,
])

export type TokenMessage = z.infer<typeof TokenMessageScheme>
export type AuthenticatedMessage = z.infer<typeof AuthenticatedMessageScheme>
export type GameMoveMessage = z.infer<typeof GameMoveMessageScheme>
export type GameStateMessage = z.infer<typeof GameStateMessageScheme>
export type GameCreatedMessage = z.infer<typeof GameCreatedMessageScheme>
export type ConnectedMessage = z.infer<typeof ConnectedMessageScheme>
export type DisconnectedMessage = z.infer<typeof DisconnectedMessageScheme>
export type InvalidMoveMessage = z.infer<typeof InvalidMoveMessageScheme>
export type GameMessage = z.infer<typeof GameMessageScheme>
